### PR TITLE
Support running all tests in project

### PR DESCRIPTION
**Pull request by jacobmendoza**
Modifications to support running all tests in project. Looking for some feedback to see if everything looks in the right direction and some implied assumptions are fine.
- The order of the files appears in the same order returned by the TAP reporter.
- I haven't removed TerminalCommandExecutor yet but this class is now using BufferedProcess. BufferedProcess produces side effects as it already spawns the process and could make testing harder as it is hard to mock. I'm trying to figure out this before completely removing the class. For now, acts as a wrapper (related: https://github.com/avajs/atom-ava/issues/16).
- As suggested, the package will execute all tests with ctrl-alt-a and the current file with ctrl-alt-shift-a. 
- An editor screen must be opened to deduce the project. 
- This extension needs AVA 0.15 to work properly (https://github.com/avajs/ava/commit/3c4babd0efed2ddd9cfe6e2dc23952338f0c9879). Previous versions will display the ANSI scape codes. I think I should be removing them also to support previous versions, not done yet.

This feature was already started, but IMO next step should be migrate the test suite to AVA. 

screen shot 2016-05-29 at 18 01 04

Related: https://github.com/avajs/atom-ava/issues/13

### CODE REVIEW

File: keymaps/ava.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
 	"atom-workspace": {
-		"ctrl-alt-r": "ava:run",
-		"ctrl-alt-a": "ava:toggle"
+		"ctrl-alt-shift-a": "ava:run",
+		"ctrl-alt-a": "ava:run-all",
+		"ctrl-alt-t": "ava:toggle"
```
Comment (sindresorhus) [inline keymaps/ava.json]: I think we should drop this shortcut. It's already covered by running. We can rather add support for pressing Esc when the pane is active to close it.

File: lib/assert-result-factory.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/** @babel */
+
+class AssertResultFactory {
```
Comment (sindresorhus) [inline lib/assert-result-factory.js]: Too much factory :p

This doesn't need to be a class. Make it a normal function instead.

File: lib/assert-result-factory.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	constructor() {
+		this.groupIsolationChar = '›';
```
Comment (sindresorhus) [inline lib/assert-result-factory.js]: You can just hardcode this. It's not going to change.

File: views/panel.html [hunk 4]
```diff
@@ -19,5 +19,5 @@
 		<div class="sk-child sk-bounce3"></div>
 	</div>
 </div>
-<div id="file-header"></div>
-<div class="tests-container"></div>
+<div id="tests-groups-container" class='tests-groups-container'>
```
Comment (sindresorhus) [inline views/panel.html]: double-quotes

File: lib/panel.js [hunk 5]
```diff
@@ -84,6 +93,14 @@ class Panel {
 		failedContainer.textContent = assertResult.currentExecution.failed;
 	}
 
+	_getCssClassForAssert(assert) {
+		if (assert.ok) {
+			return assert.skip ? 'skipped' : 'ok';
+		}
+
+		return assert.todo ? 'todo' : 'ko';
```
Comment (sindresorhus) [inline lib/panel.js]: `ko` => `ok`

File: lib/terminal-command-executor.js [hunk 6]
```diff
@@ -11,31 +11,32 @@ class TerminalCommandExecutor extends EventEmitter {
...
-			`cd \"${this.destinyFolder}\" && ${this.command}\n` :
-			`${this.command}\n`;
-
-		this.terminal.stdin.write(terminalCommand);
-		this.terminal.stdin.write('exit\n');
+		const folder = Array.isArray(destinyFolder) ?
+			destinyFolder[0] :
+			destinyFolder;
+
+		const process = new BufferedProcess({
```
Comment (sindresorhus) [inline lib/terminal-command-executor.js]: This is overwriting the global `process`. Use a different name like `cp` or `child`.

File: lib/index.js [hunk 7]
```diff
@@ -49,15 +53,34 @@ module.exports = {
...
+		if (!this.atomPanel.isVisible()) {
+			this.toggle(this._runAll);
+		}
+
+		if (!this.canRun()) {
+			return;
+		}
+
+		const editor = atom.workspace.getActiveTextEditor();
+		const currentFileName = editor.buffer.file.path;
```
Comment (sindresorhus) [inline lib/index.js]:
``` js
const currentFileName = editor.getPath();
```

File: lib/index.js [hunk 8]
```diff
@@ -49,15 +53,34 @@ module.exports = {
...
+		const paths = editor.project.rootDirectories.map(d => d.path);
+		const selectedProject = paths.filter(p => currentFileName.startsWith(p));
```
Comment (sindresorhus) [inline lib/index.js:74]:
``` js
const paths = editor.project.getPaths();
```

Instead of the two aboe lines.

File: keymaps/ava.json [hunk 9]
```diff
@@ -1,6 +1,7 @@
 {
 	"atom-workspace": {
-		"ctrl-alt-r": "ava:run",
-		"ctrl-alt-a": "ava:toggle"
+		"ctrl-alt-shift-a": "ava:run",
+		"ctrl-alt-a": "ava:run-all",
```
Comment (sindresorhus) [inline keymaps/ava.json]: I would call this: `ava:run`

File: keymaps/ava.json [hunk 10]
```diff
@@ -1,6 +1,7 @@
 {
 	"atom-workspace": {
-		"ctrl-alt-r": "ava:run",
-		"ctrl-alt-a": "ava:toggle"
+		"ctrl-alt-shift-a": "ava:run",
```
Comment (sindresorhus) [inline keymaps/ava.json]: And this: `ava:run-file`

File: lib/assert-result-factory.js (hunk 2 above)
Comment (jacobmendoza) [inline lib/assert-result-factory.js]: As always @sindresorhus, thank you so much for your feedback on this. Not sure if you're going to have time for this, but here goes my question:

Wouldn't that put too much logic in TestRunnerProcess (the client of the factory)?. Also, not 100% sure, but I'd say slightly harder to test, bigger setup needed and probably harder to understand.

Anyway, I'm already implemented the change, but would love to hear your thoughts.

File: keymaps/ava.json (hunk 1 above)
Comment (jacobmendoza) [inline keymaps/ava.json]: Researching about this. Having some problems to detect when the pane is active. Also, the part of the application that is usually active when running tests is the editor.

I'm trying to figure this out.

File: lib/index.js (hunk 8 above)
Comment (jacobmendoza) [inline lib/index.js:74]: Not sure that's equivalent.

As you said, the first line is terribly complicated and `editor.project.getPaths();` should do it. But the second line is there to infer the selected project from the currently selected file, as Atom supports different projects.

Does that make sense?. getPaths() does not give you the selected or active project.

**Comment by jacobmendoza** [comment]
Added an initial but hopefully reasonable implementation of the ESCAPE key for closing the pane. Based on: https://github.com/atom/find-and-replace/blob/master/lib/find.coffee#L78.

For our case:
- Used core:cancel. Not used core:close as the panel may not be related specifically to the current file and you want to keep it visible.
- You need to have selected the main editor, which will be active when running tests.
- Events triggered from the mini-editor (that is, other components) will be skipped.
- Toggle command and keyboard shortcut removed.

File: lib/panel.js (hunk 5 above)
Comment (jacobmendoza) [inline lib/panel.js]: Not certain about this comment. If I get it right, should be ok: 

The parser module that we're using has a flag 'ok' that will be active if the test passed. If the test is marked as todo, ok will be false and the todo flag will be true. Finally, failed tests will have ok as false, todo as false and therefore we get the 'ko' css class.

Unless I'm wrong, I think this code reflects these specs. Am I missing something?.

File: lib/assert-result-factory.js (hunk 2 above)
Comment (sindresorhus) [inline lib/assert-result-factory.js]: I'm not saying the logic should be in TestRunnerProcess. It should be here, but it would be better to just export a simple function instead of a class with one function. Having it as a class here brings no benefit.

File: lib/panel.js (hunk 5 above)
Comment (sindresorhus) [inline lib/panel.js]: Should have been clearer. I'm aware how it works, but `ko` is a bad name. Should rather be `not-ok`, and updated in the LESS file too.

File: views/panel.html [hunk 11]
```diff
@@ -19,5 +19,5 @@
 		<div class="sk-child sk-bounce3"></div>
 	</div>
 </div>
-<div id="file-header"></div>
-<div class="tests-container"></div>
+<div id='tests-groups-container' class='tests-groups-container'>
```
Comment (sindresorhus) [inline views/panel.html]: Double-quotes for HTML attributes.

File: lib/assert-result-factory.js (hunk 2 above)
Comment (jacobmendoza) [inline lib/assert-result-factory.js]: Oh, I see it now. Yeah, agreed! :).

File: lib/panel.js (hunk 5 above)
Comment (jacobmendoza) [inline lib/panel.js]: That makes sense. Will update.

**Comment by jacobmendoza** [comment]
@sindresorhus, does this look mature enough to be merged?

File: lib/assert-result-factory.js [hunk 12]
```diff
@@ -1,33 +1,22 @@
...
-		this.groupIsolationChar = '›';
-	}
-
-	getAssertResult(assert, context) {
-		const result = {
-			assert,
-			group: '',
-			currentExecution: context
-		};
+module.exports.getAssertResult = function (assert, context) {
```
Comment (sindresorhus) [inline lib/assert-result-factory.js]: Just export it directly:

``` js
module.exports = (assert, context) => {
```

Note arrow function too.

And just rename the file to `get-assert-result.js`.

**Comment by sindresorhus** [comment]
Only one thing and it's ready: https://github.com/avajs/atom-ava/pull/18#discussion_r69033888

File: lib/assert-result-factory.js (hunk 12 above)
Comment (jacobmendoza) [inline lib/assert-result-factory.js]: Oh, agreed!. Thank you so much @sindresorhus

**Comment by jacobmendoza** [comment]
@sindresorhus, as always, very grateful for all the feedback. Really appreciated.

**Comment by jacobmendoza** [comment]
Found and edge case yesterday. When the user selected 'run all the files' but only one test was in the folder, the package was unable to render the results.

On my side, ready to merge now @sindresorhus .

**Comment by sindresorhus** [comment]
Excellent. Works great! Thanks for your hard work on this :)

**Comment by jacobmendoza** [comment]
Thank you @sindresorhus for all your feedback and support. Learning lots of things!.
